### PR TITLE
Dynamic RPC Server and Node connection map

### PR DESCRIPTION
**Pull request by dadgar**
This PR introduces a dynamic RPC server per connection and adds a mapping between node IDs and connections.

### CODE REVIEW

File: nomad/node_endpoint.go [hunk 1]
```diff
@@ -305,6 +315,13 @@ func (n *Node) UpdateStatus(args *structs.NodeUpdateStatusRequest, reply *struct
 		return fmt.Errorf("node not found")
 	}
 
+	// We have a valid node connection, so add the mapping to cache the
+	// connection and allow the server to send RPCs to the client.
+	if n.ctx != nil && n.ctx.NodeID == "" {
+		n.ctx.NodeID = args.NodeID
+		n.srv.addNodeConn(n.ctx)
+	}
+
```
Comment (preetapan) [inline nomad/node_endpoint.go]: Is there a reason why only a few endpoints attempt to add a valid connection like this? for e.g why not do the same thing in GetNode?

File: nomad/rpc.go [hunk 2]
```diff
@@ -126,7 +154,37 @@ func (s *Server) handleConn(conn net.Conn, isTLS bool) {
...
+			return
+		}
+
+		// Update the connection context with the fact that the connection is
+		// using TLS
+		ctx.TLS = true
+
+		// Parse the region and role from the TLS certificate
+		state := tlsConn.ConnectionState()
+		parts := strings.SplitN(state.ServerName, ".", 3)
```
Comment (preetapan) [inline nomad/rpc.go]: Is this name guarenteed to be in this format always?

File: nomad/rpc.go [hunk 3]
```diff
@@ -126,7 +154,37 @@ func (s *Server) handleConn(conn net.Conn, isTLS bool) {
...
+		if len(parts) != 3 || (parts[0] != "server" && parts[0] != "client") || parts[2] != "nomad" {
```
Comment (preetapan) [inline nomad/rpc.go]: parts[1] != "client" here ?

File: nomad/rpc.go [hunk 4]
```diff
@@ -126,7 +154,37 @@ func (s *Server) handleConn(conn net.Conn, isTLS bool) {
...
+			s.logger.Printf("[WARN] nomad.rpc: invalid server name %q on verified TLS connection", state.ServerName)
```
Comment (preetapan) [inline nomad/rpc.go]: This validation is hard to follow, why do we care about parts[3] == "nomad" if we only ever read parts [0] and [1]

File: nomad/rpc.go (hunk 2 above)
Comment (dadgar) [reply nomad/rpc.go]: It should be in the case of verifying hostnames but it is a bit brittle. Going to rework

Comment (dadgar) [reply nomad/rpc.go]: See newest commit. I store the whole verified chain so we can work with the certificates where needed.

File: nomad/rpc.go (hunk 4 above)
Comment (dadgar) [reply nomad/rpc.go]: Removed.
